File: 백준/Silver/18353. 병사 배치하기/병사 배치하기.py
import sys
import bisect

input = sys.stdin.readline

def lower_bound(arr,target):
    n=len(arr)
    for i in range(n):
        if arr[i]<=target:
            return i
    return n

def lis(arr):
    lis=[]
    for x in arr:
        i=lower_bound(lis,x)
        if i==len(lis):
            lis.append(x)
        else:
            lis[i]=x
    return len(lis)

n=int(input())
arr=list(map(int,input().split()))

print(n-lis(arr))

File: 백준/Silver/18353. 병사 배치하기/병사 배치하기.cc
#include <bits/stdc++.h>
using namespace std;

int lis(vector<int>&arr) {
    vector<int>lis;
    for (int i:arr) {
        auto it=lower_bound(lis.begin(),lis.end(),i,greater<int>());
        if (it==lis.end()) lis.push_back(i);
        else *it=i;
    }
    return lis.size();
}

int main() {
    cin.tie(0)->sync_with_stdio(0);

    int n;
    cin>>n;
    vector<int>arr(n);
    for (int i=0; i<n; i++) {
        cin>>arr[i];
    }
    cout<<n-lis(arr);
}